bitcoin/src/blockdata/script.rs

```diff
@@ -400,6 +400,26 @@ impl Script {
       Builder::new()
     }
 
+    /// Returns the public key if this script is P2PK with a valid public key.
+    pub fn p2pk_public_key(&self) -> Option<PublicKey> {
```

Comment:
Now I'm thinking, should this have API similar to `std::env::var()`? That means return `Result<PublicKey, P2PkError>` where:

```rust
// Not `#[non_exhaustive]` - there should be no reason to add variants
pub enum P2PkError {
    NotP2Pk /* have some struct here storing more information? I guess it's too much? */,
    InvalidKey(secp256k1::Error),
}
```

What do you think @apoelstra @tcharding?

Reply:
If we went this direction, we could actually have a tri-partite result
```
enum P2pPkResult {
    NotP2Pk,
    InvalidKey([u8; 65], secp256k1::Error),
    Key(bitcoin::PublicKey),
}
```
Actually we should even split the `InvalidKey` into two variants, one with a 33-byte "compressed key" and another with a 65 byte "uncompressed key".

**But** to be honest I think the option is fine. The purpose of this function is for block explorers who just want to show the pubkey to users in the case that they've got a p2pk, and don't care why something isn't a p2pk.

Reply:
Why would we invent our own result type instead of `Result<PublicKey, P2PkError>`?

> don't care why something isn't a p2pk.

Showing that to the user would make a more interesting explorer IMO. :)

> split the `InvalidKey` into two variants, one with a 33-byte "compressed key" and another with a 65 byte "uncompressed key".

I was actually thinking about having `PublicKeyBytes` type with meaning "bytes intended to represent public key which may be invalid". I suspect it could be used in a s few more places. However I since then figured out the code is structured a bit differently than I thought.

Reply:
>Why would we invent our own result type

Because result has two variants, and we want 3 (or 4), and none are clearly errors.

Reply:
Oh, I see what you mean. I don't think it's too bad to use `Result` though. The line between errors and "something that resembles error, such as `None`" is blurry. The existence of `.ok()` method on `Result` and `ok_or()` on `Option` demonstrates that. Just as "item missing in a map" may or may not be error depending on the context it's used in, `Err` may be actually fine in many cases. For a different example I ignore `io::ErrorKind::NotFound` for missing optional configuration files.

If you think it could confuse newbies we could just point it out in the doc.

Reply:
@gd87429 would returning `Result` be useful to you here? Would it be helpful to return bytes of an invalid key?

Reply:
@Kixunil The `Option` would be more than enough for my simple usecase. Thank you for pinging.

Reply:
OK, I guess let's use `Option` unless someone asks for `Result`.